File: apps/android/app/src/main/java/fr/fuelradar/data/DeptIndex.kt
package fr.fuelradar.data

import fr.fuelradar.data.model.DeptBbox
import kotlin.math.PI
import kotlin.math.cos

/** Axis-aligned bounding box in degrees. */
data class BoundingBox(
    val minLat: Double,
    val maxLat: Double,
    val minLng: Double,
    val maxLng: Double,
)

/**
 * Pure geo helpers mirroring apps/web/src/lib/department.ts and deptIndex.ts.
 * No Android dependencies, so this is unit-testable on the JVM.
 */
object DeptIndex {

    /** Convert a 5-digit French postal code to its department code. */
    fun getDepartment(cp: String): String {
        if (cp.startsWith("97")) return cp.take(3)
        if (cp.startsWith("20")) {
            val num = cp.toIntOrNull() ?: 0
            return if (num < 20200) "2A" else "2B"
        }
        return cp.take(2)
    }

    /** Approx square covering a circle of [radiusKm] around a coordinate. */
    fun boundingBox(lat: Double, lng: Double, radiusKm: Double): BoundingBox {
        val dLat = radiusKm / 111.0
        val dLng = radiusKm / (111.0 * cos(lat * PI / 180.0))
        return BoundingBox(
            minLat = lat - dLat,
            maxLat = lat + dLat,
            minLng = lng - dLng,
            maxLng = lng + dLng,
        )
    }

    /** Department codes whose station bbox overlaps [target]. */
    fun findOverlappingDepts(index: DeptBbox, target: BoundingBox): List<String> {
        val out = mutableListOf<String>()
        for ((dept, bb) in index) {
            if (bb.size < 4) continue
            val minLat = bb[0]
            val maxLat = bb[1]
            val minLng = bb[2]
            val maxLng = bb[3]
            val overlaps =
                maxLat >= target.minLat &&
                    minLat <= target.maxLat &&
                    maxLng >= target.minLng &&
                    minLng <= target.maxLng
            if (overlaps) out.add(dept)
        }
        return out
    }

    /** Department codes whose bbox overlaps a circle of [radiusKm] around a point. */
    fun deptsAround(index: DeptBbox, lat: Double, lng: Double, radiusKm: Double): List<String> =
        findOverlappingDepts(index, boundingBox(lat, lng, radiusKm))
}

File: apps/web/src/lib/deptIndex.ts
import { boundingBox } from './department';

const BASE = import.meta.env.BASE_URL;

/** [minLat, maxLat, minLng, maxLng] per department code. */
export type DeptBbox = Record<string, [number, number, number, number]>;

let promise: Promise<DeptBbox | null> | null = null;

export async function fetchDeptBbox(): Promise<DeptBbox | null> {
  if (!promise) {
    promise = (async () => {
      try {
        const res = await fetch(`${BASE}data/dept-bbox.json`);
        if (!res.ok) return null;
        const ct = res.headers.get('content-type') ?? '';
        if (!ct.includes('json')) return null;
        return (await res.json()) as DeptBbox;
      } catch {
        return null;
      }
    })();
  }
  return promise;
}

/**
 * Dept codes (French departments + ES/PT pseudo-departments) whose
 * station bbox overlaps a circle of radiusKm around a point. Returns null
 * when the index could not be loaded (caller should fall back).
 */
export async function deptsAround(
  lat: number,
  lng: number,
  radiusKm: number,
): Promise<string[] | null> {
  const index = await fetchDeptBbox();
  if (!index) return null;
  return findOverlappingDepts(index, boundingBox(lat, lng, radiusKm));
}

/** Returns dept codes whose station bbox overlaps a target bounding box. */
export function findOverlappingDepts(
  index: DeptBbox,
  target: { minLat: number; maxLat: number; minLng: number; maxLng: number },
): string[] {
  const out: string[] = [];
  for (const [dept, bb] of Object.entries(index)) {
    const [minLat, maxLat, minLng, maxLng] = bb;
    const overlaps =
      maxLat >= target.minLat &&
      minLat <= target.maxLat &&
      maxLng >= target.minLng &&
      minLng <= target.maxLng;
    if (overlaps) out.push(dept);
  }
  return out;
}
